## Soru 1

* Harcamalar hedef değişkenimizdir
* Yıllık gelir, karesi ve çocuk sahibi olma özellik değişkenleridir.
* Yıllık gelir arttıkça ax^2 + bx şeklinde bir artış görürüz.
* Çocuk sahibi olmak harcamaları negatif yönde etkilemektedir.
* İstatitistiksel olarak anlamlı olup olmadığına değişkenlerin p değerlerine bakarak yorum yapabiliriz.


## Soru 2 

Bu alıştırmada, önceki dersteki sıcaklık verileriyle çalışacaksınız. Bu ödevi tamamlamak için, gist dosyasına, aşağıdaki görevlere ilişkin çözümlerinizi içeren Jupyter notebook'u bir link olarak ekleyin:

İlk olarak, Sıcaklık verilerini Kaggle'dan indirin.
Hedef değişkeninizin görünür sıcaklık ve sıcaklık arasındaki fark olduğu doğrusal bir regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. Şimdi, modelinizi OLS kullanarak tahmin edin. Tahmin edilen katsayılar istatistiksel olarak anlamlı mıdır? Tahmini katsayılar önceki beklentileriniz doğrultusunda mı? Tahmin edilen katsayıları yorumlayınız. Hedef ve açıklayıcı değişkenler arasındaki ilişkiler nelerdir?
Ardından, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Katsayılar istatistiksel olarak anlamlı mıdır? Nem ve rüzgar hızı için tahmini katsayıların işaretleri değişti mi? Tahmin edilen katsayıları yorumlayınız.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import scipy.stats as sts
import statsmodels.api as sm
import warnings


from statsmodels.tsa.stattools import acf
from sklearn.preprocessing import normalize
from sklearn.preprocessing import scale
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from scipy.stats.mstats import winsorize
from scipy.stats import jarque_bera
from scipy.stats import normaltest
from scipy.stats import bartlett
from scipy.stats import levene
from sqlalchemy import create_engine
from sklearn import linear_model

warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('weatherHis.csv')
df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [3]:
temp = df['Temperature (C)'] - df['Apparent Temperature (C)']

In [4]:
X = df[['Humidity','Wind Speed (km/h)']]
X
Y = temp

X = sm.add_constant(X)

results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sat, 19 Oct 2019   Prob (F-statistic):               0.00
Time:                        20:27:48   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

temp = -2.4381 + 3.0292Humidty + 0.1193WindSpeed

* istatiksel olarak anlamlı özellik değişkenlerimiz. P değerleri sıfır.
* Nem 1 derece arttığında sıcaklık farkı 3.0292 oranında artıyor.
* Her ikisi de artı yönde artırıyor.



In [5]:
hum_wind = df['Humidity'] * df['Wind Speed (km/h)']
df['hum_wind'] = hum_wind

X2 = df[['Humidity', 'Wind Speed (km/h)', 'hum_wind']]
Y2 = temp

X2 = sm.add_constant(X2)

results2 = sm.OLS(Y2,X2).fit()
results2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sat, 19 Oct 2019   Prob (F-statistic):               0.00
Time:                        20:32:09   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
hum_wind              0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

yeni modelde 

temp = -0.0839 -0.1775Humidty - 0.0905WindSpeed + 0.2971humwind

* nem ve rüzgar eksi yöne döndüler
* sabit azaldı
* p değerleri yine 0.05'ten küçük, istatistiksel olarak anlamlılar.
* yeni özellik değişkeni daha büyük bir etkiye sahip oldu ve artı yönde.


## Soru 3

Bu alıştırmada, ev fiyatları modelinizi yorumlayacaksınız. Bu ödevi tamamlamak için, gist dosyasına, aşağıdaki görevlere ilişkin çözümlerinizi içeren Jupyter notebook'u bir link olarak ekleyin:

Ev Fiyatları verilerini Kaggle'dan indirin.
Ev fiyatları modelinizi tekrar çalıştırın ve sonuçları yorumlayın. Hangi özellikler istatistiksel olarak anlamlı ve hangileri değildir?
Şimdi, anlamsız özellikleri modelinizden hariç tutun. Bir şey değişti mi?
Ev fiyatları ile olan ilişkilerini belirleyerek istatistiksel olarak anlamlı katsayıları yorumlayın. Ev fiyatları üzerinde hangi özellikler daha belirgin bir etkiye sahiptir?
Sonuçlar size mantıklı geliyor mu? Değilse, potansiyel nedenleri açıklamaya çalışın.

In [6]:
df = pd.read_csv("trainh.csv")
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [7]:
df = df.drop([ 'PoolQC','MiscFeature', 'Alley','Fence','FireplaceQu', 'PoolQC', 'MiscFeature'], axis = 1)

In [8]:
df.dropna(inplace=True)

In [9]:
Y=df["SalePrice"]
X=df[["OverallQual", "TotalBsmtSF", "1stFlrSF", "GrLivArea", 'GarageCars', 'GarageArea']]

regr=linear_model.LinearRegression()
regr.fit(X, Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [10]:
X=sm.add_constant(X)
sonuc=sm.OLS(Y, X).fit()
sonuc.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.754
Model:                            OLS   Adj. R-squared:                  0.753
Method:                 Least Squares   F-statistic:                     556.1
Date:                Sat, 19 Oct 2019   Prob (F-statistic):               0.00
Time:                        20:38:07   Log-Likelihood:                -13178.
No. Observations:                1094   AIC:                         2.637e+04
Df Residuals:                    1087   BIC:                         2.640e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.226e+05   6044.073    -20.280      0.000   -1.34e+05   -1.11e+05
OverallQual  2.639e+04   1334.634     19.774      0.000    2.38e+04     2.9e+04
TotalBsmtSF    25.8991      7.475      3.465      0.001      11.233      40.565
1stFlrSF        6.1697      8.137      0.758      0.448      -9.797      22.136
GrLivArea      43.0658      3.308     13.020      0.000      36.576      49.556
GarageCars   1.823e+04   3697.570      4.929      0.000     1.1e+04    2.55e+04
GarageArea     17.1556     12.583      1.363      0.173      -7.534      41.845
==============================================================================
Omnibus:                      361.074   Durbin-Watson:                   1.945
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            29219.745
Skew:                          -0.556   Prob(JB):                         0.00
Kurtosis:                      28.294   Cond. No.                     1.16e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.16e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [11]:
print("SalePrice= {:.3f} +\n {:.3f} OverallQual +\n {:.3f} TotalBsmtSF +\n {:.3f} 1stFlrSF +\n {:.3f} GrLivArea +\n {:.3f} GarageCars +\n {:.3f}  GarageArea +\n"
      .format(regr.intercept_, regr.coef_[0], regr.coef_[1], regr.coef_[2], regr.coef_[3], regr.coef_[4], regr.coef_[5]))

SalePrice= -122571.937 +
 26390.678 OverallQual +
 25.899 TotalBsmtSF +
 6.170 1stFlrSF +
 43.066 GrLivArea +
 18226.869 GarageCars +
 17.156  GarageArea +



* 1stFlrSF ve GarageArea istatistiksel olarak anlamlı değiller, p değerleri yüksek


In [12]:
Y=df["SalePrice"]
X=df[["OverallQual", "TotalBsmtSF", "GrLivArea", 'GarageCars']]

regr=linear_model.LinearRegression()
regr.fit(X, Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [13]:
X=sm.add_constant(X)
sonuc=sm.OLS(Y, X).fit()
sonuc.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.754
Model:                            OLS   Adj. R-squared:                  0.753
Method:                 Least Squares   F-statistic:                     833.1
Date:                Sat, 19 Oct 2019   Prob (F-statistic):               0.00
Time:                        20:40:50   Log-Likelihood:                -13179.
No. Observations:                1094   AIC:                         2.637e+04
Df Residuals:                    1089   BIC:                         2.639e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.215e+05   5858.583    -20.736      0.000   -1.33e+05    -1.1e+05
OverallQual  2.628e+04   1331.186     19.739      0.000    2.37e+04    2.89e+04
TotalBsmtSF    32.1067      3.767      8.523      0.000      24.715      39.498
GrLivArea      44.1656      3.123     14.143      0.000      38.038      50.293
GarageCars   2.195e+04   2483.929      8.837      0.000    1.71e+04    2.68e+04
==============================================================================
Omnibus:                      358.560   Durbin-Watson:                   1.945
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            28329.981
Skew:                          -0.552   Prob(JB):                         0.00
Kurtosis:                      27.905   Cond. No.                     9.36e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.36e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [14]:
print("SalePrice= {:.3f} +\n {:.3f} OverallQual +\n {:.3f} TotalBsmtSF +\n  {:.3f} GrLivArea +\n {:.3f} GarageCars +\n "
      .format(regr.intercept_, regr.coef_[0], regr.coef_[1], regr.coef_[2], regr.coef_[3]))

SalePrice= -121482.925 +
 26276.362 OverallQual +
 32.107 TotalBsmtSF +
  44.166 GrLivArea +
 21949.307 GarageCars +
 


* değişkenlerimizin kat sayıları değişti ancak aşırı kritik bir değişiklik de bulunmuyor, yönler değişmedi
* En etkili olan OverallQual, hala en etkili özellik değişkeni olarak kalmaya devam etti.
* tüm p değerleri düşük, istatistiksel olarak anlamlı